neqo-common/src/datagram.rs

```diff
@@ -7,20 +7,129 @@
+    Af13 = 0b0011_1000,       // Assured Forwarding 13, [RFC2597]
+    Af21 = 0b0100_1000,       // Assured Forwarding 21, [RFC2597]
+    Af22 = 0b0101_0000,       // Assured Forwarding 22, [RFC2597]
+    Af23 = 0b0101_1000,       // Assured Forwarding 23, [RFC2597]
+    Af31 = 0b0110_1000,       // Assured Forwarding 31, [RFC2597]
+    Af32 = 0b0111_0000,       // Assured Forwarding 32, [RFC2597]
+    Af33 = 0b0111_1000,       // Assured Forwarding 33, [RFC2597]
+    Af41 = 0b1000_1000,       // Assured Forwarding 41, [RFC2597]
+    Af42 = 0b1001_0000,       // Assured Forwarding 42, [RFC2597]
+    Af43 = 0b1001_1000,       // Assured Forwarding 43, [RFC2597]
+    Ef = 0b1011_1000,         // Expedited Forwarding, [RFC3246]
+    VoiceAdmit = 0b1011_0000, // Capacity-Admitted Traffic, [RFC5865]
+    Le = 0b0000_0100,         // Lower-Effort, [RFC8622]
+}
+
+impl std::fmt::Debug for IpTosDscp {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            IpTosDscp::Cs0 => f.write_str("CS0"),
+            IpTosDscp::Cs1 => f.write_str("CS1"),
+            IpTosDscp::Cs2 => f.write_str("CS2"),
+            IpTosDscp::Cs3 => f.write_str("CS3"),
+            IpTosDscp::Cs4 => f.write_str("CS4"),
+            IpTosDscp::Cs5 => f.write_str("CS5"),
+            IpTosDscp::Cs6 => f.write_str("CS6"),
+            IpTosDscp::Cs7 => f.write_str("CS7"),
+            IpTosDscp::Af11 => f.write_str("AF11"),
+            IpTosDscp::Af12 => f.write_str("AF12"),
+            IpTosDscp::Af13 => f.write_str("AF13"),
+            IpTosDscp::Af21 => f.write_str("AF21"),
+            IpTosDscp::Af22 => f.write_str("AF22"),
+            IpTosDscp::Af23 => f.write_str("AF23"),
+            IpTosDscp::Af31 => f.write_str("AF31"),
+            IpTosDscp::Af32 => f.write_str("AF32"),
+            IpTosDscp::Af33 => f.write_str("AF33"),
+            IpTosDscp::Af41 => f.write_str("AF41"),
+            IpTosDscp::Af42 => f.write_str("AF42"),
+            IpTosDscp::Af43 => f.write_str("AF43"),
+            IpTosDscp::Ef => f.write_str("EF"),
+            IpTosDscp::VoiceAdmit => f.write_str("VOICE-ADMIT"),
+            IpTosDscp::Le => f.write_str("LE"),
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Default)]
+pub struct IpTos((IpTosDscp, IpTosEcn));
+
+impl std::fmt::Debug for IpTos {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_tuple("IpTos")
+            .field(&self.0 .0)
+            .field(&self.0 .1)
+            .finish()
+    }
+}
+
+#[derive(Clone, PartialEq, Eq)]
 pub struct Datagram {
     src: SocketAddr,
     dst: SocketAddr,
+    tos: IpTos,
+    ttl: Option<u8>,
     d: Vec<u8>,
 }
 
 impl Datagram {
-    pub fn new<V: Into<Vec<u8>>>(src: SocketAddr, dst: SocketAddr, d: V) -> Self {
+    pub fn new<V: Into<Vec<u8>>>(
+        src: SocketAddr,
+        dst: SocketAddr,
+        tos: IpTos,
+        ttl: Option<u8>,
```

Comment:
FWIW, you could also new-type `IpTtl(u8)` and `impl Default for IpTtl { fn default() -> Self { Self(128) } }` to save an `Option<>` wrapper.

I guess that the only potential problem is that if the OS has some different default for TTL, we might start overriding that unnecessarily.  I'm not sure that I care about that though.